runtime/j9vm/jvm.c

```diff
@@ -3989,15 +3989,49 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
 			UDATA slOpenResult = j9sl_open_shared_library((char *)libName, &handle, flags);
 			Trc_SC_LoadLibrary_OpenShared(libName);
 
-/* jdk17+ calls JVM_LoadLibrary with decorated library names. If the following is done
- * then it overwrites the real error with a failure to load "liblib<name>.so.so".
- */
-#if JAVA_SPEC_VERSION < 17
 			if (0 != slOpenResult) {
-				slOpenResult = j9sl_open_shared_library((char *)libName, &handle, flags | J9PORT_SLOPEN_DECORATE);
-				Trc_SC_LoadLibrary_OpenShared_Decorate(libName);
-			}
+#if JAVA_SPEC_VERSION >= 17
+				/* JDK17+ calls JVM_LoadLibrary() with decorated library names.
+				 * The library name passed to j9sl_open_shared_library() with the flag J9PORT_SLOPEN_DECORATE
+				 * has to be platform independent, i.e., must not contain any prefix or file extension.
+				 * A path is supported.
+				 */
+				const char *fileExt = strrchr(libName, '.');
+				BOOLEAN doOpenLibrary = TRUE;
+				if (NULL != fileExt) {
+					/*  strlen("lib") = 3 */
+					const size_t libStrLength = 3;
+					const size_t libSuffixLength = strlen(fileExt);
+					const char *fileName = strrchr(libName, '/');
+					char libPath[EsMaxPath] = {0};
```

Comment:
This makes unwarranted assumptions:
* The first three characters of the final segment of the path may not be "lib"; this should not replace a request for "foobar.so" with "bar".
* On Windows, there is no such prefix, and so one should not be removed.
* On Windows, the directory separator is `\`, not `/`.
* Likewise, a suffix should only be removed if it is the expected extension.

This should not impose an arbitrary length limit on the path name.

A pointer to `libPath` cannot be use beyond the scope of that variable; at the end of the scope, the compiler is free to reuse that storage in any way it sees fit.

Reply:
* Added support for `WIN32` and `J9OS_I5` which don't have `lib` prefixes;
* Use `DIR_SEPARATOR` for directory separator searching;
* The prefix/suffix are added via `Java_java_lang_System_mapLibraryName()` within  [jdk.internal.loader.NativeLibraries](https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/0b87eb4e968b8abe041eb9e551b3e9f7ffdba045/src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java#L277), the appropriate platform-specific suffix is expected;
* `EsMaxPath` limit is in [j9sl_open_shared_library()](https://github.com/eclipse-openj9/openj9-omr/blob/de311dd2b79515a5c8472b292f527314ce4f4c21/port/aix/omrsl.c#L238) which is invoked by `JVM_LoadLibrary()` here;
* Created a temporary local `char*` pointer for `libPath` to avoid the out-of-scope `libName`.

Reply:
This doesn't address the first point: a failed attempt to load `foobar.so` should not be satisfied by loading `libbar.a` (on AIX), for example. This should not assume that `NativeLibraries` is the only user of this function.

The length limit imposed by OMR is `MAX_STRING_LENGTH` which may or may not be the same as `EsMaxPath`: This should not make assumptions about internally defined limits in OMR.

`omrsl.c` imposes no limit when `J9PORT_SLOPEN_DECORATE` is not specified, so it seems unreasonable that a path longer than the limit imposed by OMR which could be opened if `JVM_LoadLibrary()` received that path directly could not be opened if the suffix change is appropriate. Suppose `/very/long.../` is a prefix that makes both paths below longer than both `EsMaxPath` and `MAX_STRING_LENGTH` and that the `.a` file exists and is accessible. The call `JVM_LoadLibrary("/very/long/.../libsample.a")` would succeed, but the call `JVM_LoadLibrary("/very/long/.../libsample.so")` would fail (because the path is considered too long). Loading a library is a relatively infrequent operation, so the possible expense of allocating a sufficiently large buffer is reasonable.

Reply:
* Check the library name prefix for those callers not like `NativeLibraries.load()`;
* Dynamically allocate memory for the library path longer than `EsMaxPath`, the file name length is expected to be contained in `libPath`.